### ISSUE TITLE
Dataset prefetch not working as expected, not storing data in memory

### ISSUE THREAD

**Issue by mimxrt** [tensorflow/tensorflow#31616]
**System information**
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04 LTS
- **TensorFlow installed from (source or binary)**: conda-forge
- **TensorFlow version (use command below)**: unknown 1.14.0
- **Python version**: Python 3.7.3
- **CUDA/cuDNN version**: NVIDIA-SMI 418.67, Driver Version: 418.67, CUDA Version: 10.1
- **GPU model and memory**: Quadro RTX 6000, 24190MiB
- **Exact command to reproduce**:

**Describe the current behavior**

I am training a small LSTM model and until recently, I could use `Dataset.from_tensor_slices()` reading numpy arrays directly because all training data fits into memory. Unfortunately, after adding some new data, I ran into the 2 GB graph memory limitation and was forced to switch to using `TFRecord` and `TFRecordDataset`. However, the actual training data still fits into RAM and I want to make sure it is prefetched even when using the `TFRecordDataset`. Therefore, I tried to use the `Dataset.prefetch()` methodology to achieve this, assuming a buffer will be created and (constantly) filled with data. However, it does not work - in fact, there seems to be little to no difference comparing a version with and without a final `.prefetch(x)` in the data pipeline. See the animated gif below:

![tf_prefetch](https://user-images.githubusercontent.com/53339396/63017447-a985c980-be96-11e9-91ce-0ea3cd23945f.gif)

The actual dataset is filtered in the pipeline and the training stalls whenever a sequence of values that is filtered out is occurring in the data. Only a few values in each data/tfrecord file (of which many exist) are relevant. To illustrate this further, the data layout is similar to this:

    file 1: [---------#####----------------####------]
    file 2: [--------######----------------###-------]
    file 3: [----------####----------------####------]
    file 4: ...
    ...

where `-` denotes irrelevant and `#` denotes relevant data points in a time series. When holding all values in memory (as previously was the case), the filter is rather fast and irrelevant values are skipped unnoticeable.

The data pipeline is set up like this:

    feature_description = { "features": tf.FixedLenFeature([132], tf.float32),  "label": tf.FixedLenFeature([1], tf.float32) }
    def _parse_function(example_proto):
        return tf.parse_single_example(example_proto, feature_description)
    
    ds = tf.data.TFRecordDataset([f.as_posix() for f in fs_train])
    ds = ds.map(_parse_function)
    ds = ds.flat_map(lambda v: tf.data.Dataset.from_tensors((v["features"][2:], v["label"])))
    # filter data, only allow Ls[0] and Ls[1]
    ds = ds.filter(
        lambda _, y: tf.reshape(tf.logical_or(
            tf.equal(y, Ls[0]),
            tf.equal(y, Ls[1])
        ), [])
    )
    # relabel and re-map labels to 0 and 1
    ds = ds.flat_map(lambda x, y: tf.data.Dataset.from_tensors((x, tf_relabel(y) - base_label.value)))
    # create sliding window for LSTM
    ds = ds.window(size=window_size, shift=shift, stride=stride, drop_remainder=True)
    ds = ds.flat_map(lambda x, y: tf.data.Dataset.zip((x.batch(window_size), y.batch(window_size))))
    
    # batch and prefetch
    ds = ds.batch(batch_size, drop_remainder=True)
    ds = ds.prefetch(1000000000000000) # tried many values, nothing works

**Describe the expected behavior**
I expect to find some value for `Dataset.prefetch()` that reads all or enough data to memory to allow for fast training without stalling.

**Code to reproduce the issue**
See data pipeline above. I cannot provide the data as it is proprietary.

**Comment by srihari-humbarwadi** [comment]
@mimxrt try using the
`tf.data.Dataset.cache` method
```
**Args**:
filename: A tf.string scalar tf.Tensor, representing the 
name of a directory on the filesystem to use for caching 
tensors in this Dataset. If a filename is not provided, 
the dataset will be cached in memory.
```
Note : Its is good practice to cache the dataset before performing any non deterministic operations on it, (like random augmentations)

**Comment by mimxrt** [comment]
Thank you for your response. From my understanding, while the `cache()` method would work from the second epoch forward, it does not change the behavior for the first epoch at all. So while I will certainly make use of `cache()` in future, my expectation is still that `prefetch()` reads data to memory _before_ any operation is done on the data. My reason for this is that before switching to `TFRecordDataset` (i.e. when using `.from_tensor_slices()` on Numpy arrays), the first epoch was running as fast as it can get and the whole training of 50 epochs was finished in under one hour. Now it takes over one hour just for the first epoch. The difference was that previously the raw data was sequentially read into memory as-is, without any operation on it.

It seems there should be a way to cache the dataset before any further processing; if not `preprocess()` then what else? NB: I also tried to put `preprocess()` to an earlier stage of the data pipeline.

**Comment by srihari-humbarwadi** [comment]
@mimxrt if you are purely aiming for throughput, you might be better off doing offline filtering.
Also there are a couple of low hanging fruits viz.
 - `num_parallel_calls` arg in `map` for parallel processing of dataset elements
-  `tf.data.experimental.parallel_interleave` or `tf.data.Dataset.interleave `when processing tfrecords. However parallel_interleave is deprecated.
- `tf.data.experimental.AUTOTUNE` for `buffer_size ` in `prefetch` and `num_parallel_calls` in `map`
- Tune the pipeline according to your needs using `tf.data.Options`

**Comment by mimxrt** [comment]
I understand, thank you for your kind help. I will try your suggestions and then decide to do offline filtering if necessary.

**Comment by mimxrt** [comment]
I'm sorry to re-open this but it seems again that both `cache()` and `num_parallel_calls` have no effect. See the screenshot below for the second epoch when using `cache()` after pre-processing, i.e. before windowing with `window()`.

![2019-08-15-123156_1920x1080_scrot](https://user-images.githubusercontent.com/53339396/63089655-dac6ce00-bf58-11e9-8a46-ba62b8e4291e.png)

Before, when using Numpy array input, the very same network was trained like this:

![2019-08-15-124847_1920x1080_scrot](https://user-images.githubusercontent.com/53339396/63090289-1e223c00-bf5b-11e9-8e9d-add9756feb5a.png)

For comparison: 145 s / 27485 items = 5.28 ms per item, versus now ~81 ms per item.

So for the run above, the pipeline looks like this:

```
ds_train = tf.data.TFRecordDataset([f.as_posix() for f in fs_train])
#ds_train = ds_train.cache()

ds_train = ds_train.map(lambda x: tf.parse_single_example(x, {
    "features": tf.FixedLenFeature([132], tf.float32),
    "label": tf.FixedLenFeature([1], tf.float32),
}))

ds_train = ds_train.flat_map(lambda v: tf.data.Dataset.from_tensors((v["features"][2:], v["label"])))

ds_train = ds_train.filter(
    lambda _, y: tf.reshape(tf.logical_or(
        tf.equal(y, Ls[0]),
        tf.equal(y, Ls[1])
    ), [])
)

ds_train = ds_train.flat_map(lambda x, y: tf.data.Dataset.from_tensors((x, y - base_label.value)))

ds_train = ds_train.cache()

ds_train = ds_train.window(size=num_tsteps, shift=1, stride=1, drop_remainder=True)
ds_train = ds_train.flat_map(lambda x, y: tf.data.Dataset.zip((x.batch(num_tsteps), y.batch(num_tsteps))))

ds_train = ds_train.batch(batch_size, drop_remainder=True)
ds_train = ds_train.prefetch(1)
```

I tried to add the suggested optimizations with this pipeline:
```
ds_train = tf.data.TFRecordDataset([f.as_posix() for f in fs_train])
#ds_train = ds_train.cache()

ds_train = ds_train.map(lambda x: tf.parse_single_example(x, {
    "features": tf.FixedLenFeature([132], tf.float32),
    "label": tf.FixedLenFeature([1], tf.float32),
}), num_parallel_calls=num_cores)

ds_train = ds_train.interleave(lambda v: tf.data.Dataset.from_tensors((v["features"][2:], v["label"])),
                               cycle_length=num_cores, block_length=1,
                               num_parallel_calls=num_cores)

ds_train = ds_train.filter(
    lambda _, y: tf.reshape(tf.logical_or(
        tf.equal(y, Ls[0]),
        tf.equal(y, Ls[1])
    ), [])
)

ds_train = ds_train.interleave(lambda x, y: tf.data.Dataset.from_tensors((x, y - base_label.value)),
                               cycle_length=num_cores, block_length=1,
                               num_parallel_calls=num_cores)

ds_train = ds_train.window(size=num_tsteps, shift=1, stride=1, drop_remainder=True)
ds_train = ds_train.interleave(lambda x, y: tf.data.Dataset.zip((x.batch(num_tsteps), y.batch(num_tsteps))),
                               cycle_length=num_cores, block_length=1,
                               num_parallel_calls=num_cores)

ds_train = ds_train.batch(batch_size, drop_remainder=True)

ds_train = ds_train.cache()
ds_train = ds_train.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
```

Running result:
![2019-08-15-122121_1920x1080_scrot](https://user-images.githubusercontent.com/53339396/63089956-f8486780-bf59-11e9-8a97-7f2e9f8c15b9.png)

NB: I put `cache()` more towards the end as it had no effect when putting it before `window()` even though it seems redundant to store windowed results...

**Comment by mimxrt** [comment]
Thank you for your comment. I understand and I am currently (hopefully) profiling the model fitting to confirm your hypothesis. Until now I am struggling to get TensorBoard to visualize the profile. Anyway, I wanted to let you know that I am still looking into this.

As a side note: `Dataset.cache()` seems to work when using a hard disk file instead of RAM. The second epoch forward is now as fast as I would expect which further supports your hypothesis.

**Comment by mimxrt** [comment]
@jsimsa Thanks again for your comment. I attached ([log_20190904150036.tar.gz](https://github.com/tensorflow/tensorflow/files/3575133/log_20190904150036.tar.gz)) a trace from the following pipeline:

    ds = tf.data.TFRecordDataset([f.as_posix() for f in files])
    ds = ds.map(lambda x: tf.parse_single_example(x, {
        "features": tf.FixedLenFeature([132], tf.float32),
        "label": tf.FixedLenFeature([1], tf.float32),
    }), num_parallel_calls=tf.data.experimental.AUTOTUNE)
    ds = ds.flat_map(lambda v: tf.data.Dataset.from_tensors((v["features"][2:], v["label"])))

    # only allow two labels
    ds = ds.filter(
        lambda _, y: tf.reshape(tf.logical_or(
            tf.equal(y, Ls[0]),
            tf.equal(y, Ls[1])
        ), [])
    )

    ds = ds.flat_map(lambda x, y: tf.data.Dataset.from_tensors((x, y - base_label.value)))
    ds = ds.cache(cachefile)
    ds = ds.window(size=num_tsteps, shift=1, stride=1, drop_remainder=True)
    ds = ds.flat_map(lambda x, y: tf.data.Dataset.zip((x.batch(num_tsteps), y.batch(num_tsteps))))
    ds = ds.batch(batch_size, drop_remainder=True)
    ds = ds.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)

I am looking forward to your insights regarding the performance. 

Additionally, I am currently running another training with a modified pipeline that (hopefully) is functionally equivalent but should be fully parallelized. However, the performance seems not to increase at all (also the CPU usage does not increase, no matter the amount of `num_cores`). Here is the trace:  [log_20190904151101.tar.gz](https://github.com/tensorflow/tensorflow/files/3575226/log_20190904151101.tar.gz)

    num_cores = 20
    ds = tf.data.TFRecordDataset([f.as_posix() for f in files])
    ds = ds.map(lambda x: tf.io.parse_single_example(x, {
        "features": tf.io.FixedLenFeature([132], tf.float32),
        "label": tf.io.FixedLenFeature([1], tf.float32),
    }), num_parallel_calls=num_cores)
    ds = ds.map(lambda v: (v["features"][2:], v["label"]), num_parallel_calls=num_cores)

    def flt(x, y):
        c = tf.cond(tf.reshape(tf.logical_or(tf.equal(y, Ls[0]), tf.equal(y, Ls[1])), []),
                    lambda: tf.constant(1, dtype=tf.int64),
                    lambda: tf.constant(0, dtype=tf.int64))
        return tf.data.Dataset.from_tensors((x, y)).repeat(c)
    ds = ds.interleave(flt, cycle_length=num_cores, block_length=1,
                      num_parallel_calls=num_cores)

    ds = ds.map(lambda x, y: (x, y - base_label.value), num_parallel_calls=num_cores)
    ds = ds.cache(cachefile)
    ds = ds.window(size=num_tsteps, shift=1, stride=1, drop_remainder=True)
    ds = ds.interleave(lambda x, y: tf.data.Dataset.zip((x.batch(num_tsteps), y.batch(num_tsteps))),
                       cycle_length=num_cores, block_length=1,
                       num_parallel_calls=num_cores)
    ds = ds.batch(batch_size, drop_remainder=True)
    ds = ds.prefetch(1)

**Comment by jsimsa** [comment]
I looked at your trace but in the five second window it captures only see one tf.data "get_next" event, which takes <40us. This suggests you are not input pipeline bound. Have you tried benchmarking the latency of your input pipeline (i.e. how long it takes to extra an element running in tight loop):

```
import timeit

start = timeit.default_timer()
i = 0
for elem in dataset:
  i += 1
end = timeit.default_timer()
print("Average latency {}".format((end - start) / i))
```
Separately from the above, try collecting a longer trace (e.g. 30 seconds).

**Comment by mimxrt** [comment]
When running your code snipped my workstation does run out of memory (128 GB RAM). For some reason, iterating the dataset like this occupies all memory after approximately 9920000 iterations (i added a `print(i)` every 10000 rounds to see some progress). After that the execution stalls and nothing happens further. However, I can confirm that the execution takes a long time (similar to running the actual training) and it seems even without further processing the generated elements (i.e. training), the pipeline is slow.

I tried to do a longer trace as well but unfortunately, the documentation you posted does only really show how to trace exactly one batch (which is probably what you see in the traces above):

```
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1, profile_batch = 3)
model.fit(train_data,
          steps_per_epoch=20,
          epochs=5, 
          callbacks=[tensorboard_callback])
```
Further down the [profiler service is briefly introduced](https://www.tensorflow.org/tensorboard/r2/tensorboard_profiling_keras#profiler_service) but `tf.python.eager.profiler.start_profiler_server()` does neither exist in TensorFlow 1.14 nor in TensorFlow 2.0.0-rc0. It seems to be the only "documented" way to use the Keras API for training (as opposed to using `session.run()` directly) and creating traces of specific length. Could you provide further instructions on how to create the trace you suggested collecting?

**Comment by jsimsa** [comment]
Have you tried using the [Profiler Service](https://www.tensorflow.org/tensorboard/r2/tensorboard_profiling_keras#profiler_service)?

**Comment by mimxrt** [comment]
Yes, sorry, I got some errors before. I now captured 60 seconds to ensure getting multiple phases of valid and invalid samples (see https://github.com/tensorflow/tensorflow/issues/31616#issue-480636359, data is only partially valid in one time series file). 

Here is the trace: [log_20190906075644.tar.gz](https://github.com/tensorflow/tensorflow/files/3582909/log_20190906075644.tar.gz)

**Comment by jsimsa** [comment]
Hrm, it is odd that the different samples look different (and the information in the trace does not seem coherent). I could try running your input pipeline myself and analyze it for you but for that, I would need end-to-end reproducible example (include the data).

**Comment by mimxrt** [comment]
Thank you for your kind offer, that's really nice. However, due to project deadlines, I had to switch to pre-processing the data in Dask and storing already the pre-processed data to tfrecord files. Doing this, the Tensorflow Dataset-pipeline is reduced to merely reading the files and the training is accelerated. Of course, I am still interested in using the more elegant way of pre-processing the data in parallel on-the-fly, but I would feel bad if you invested a lot of your time into this while there is no emergency.

Additionally, the data is proprietary, although, of course, I could provide you with fake data of the same shape and size. So, if you are interested in this as well, let me know and I will create a minimal working example including fake data. In either case, thank you very much for your kind support!

**Comment by jsimsa** [comment]
Makes sense. I would be happy to take a look at your MRE.

**Comment by Anhaoxu** [comment]
@jsimsa @srihari-humbarwadi @jvishnuvardhan @ravikyram 
Hi , I'm not familiar with tf.data api，in my case, I found the IO is a problem: GPU uti is low, about 15%. I use a generator function to read local multiple pickle files(about 200K), examples are bellow: 
```
def train_data():
    while True:
        for i in shuffle(train_id):
            df = pd.read_pickle(df_train_val['file_name'].iloc[i],
                                compression='gzip')
            feature = df.to_numpy().reshape(df.shape[0], -1)
            yield feature.reshape(-1, 56),feature.reshape(-1, 56)
```
Then
```
gn = tf.data.Dataset.from_generator(train_data, output_types=(tf.float32, tf.float32),
                                          output_shapes=((None, 56),(None, 56))
                                         ).prefetch(tf.data.experimental.AUTOTUNE)
```
I found prefetch not work and IO bottleneck is very high, so any advice for this case, how can I speed up the training process?

**Comment by jvishnuvardhan** [comment]
@Anhaoxu Please create a new issue and provide a simple standalone code to reproduce the issue. Thanks!

**Comment by jsimsa** [comment]
Please see https://github.com/tensorflow/tensorflow/issues/43905#issuecomment-823675760
